streams/test-utils/src/test/java/org/apache/kafka/streams/test/TopologyTestDriverTest.java

```diff
@@ -0,0 +1,573 @@
+
+    private TopologyTestDriver testDriver;
+    private final Properties config = new Properties() {
+        {
+            put(StreamsConfig.APPLICATION_ID_CONFIG, "test-TopologyTestDriver");
+            put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "dummy:1234");
+            put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getAbsolutePath());
+        }
+    };
+
+    private final static class Record {
+        private Object key;
+        private Object value;
+        private long timestamp;
+        private long offset;
+        private String topic;
+
+        Record(final ConsumerRecord consumerRecord) {
+            key = consumerRecord.key();
+            value = consumerRecord.value();
+            timestamp = consumerRecord.timestamp();
+            offset = consumerRecord.offset();
+            topic = consumerRecord.topic();
+        }
+
+        Record(final Object key,
+               final Object value,
+               final long timestamp,
+               final long offset,
+               final String topic) {
+            this.key = key;
+            this.value = value;
+            this.timestamp = timestamp;
+            this.offset = offset;
+            this.topic = topic;
+        }
+
+        @Override
+        public String toString() {
+            return "key: " + key + ", value: " + value + ", timestamp: " + timestamp + ", offset: " + offset + ", topic: " + topic;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final Record record = (Record) o;
+            return timestamp == record.timestamp &&
+                offset == record.offset &&
+                Objects.equals(key, record.key) &&
+                Objects.equals(value, record.value) &&
+                Objects.equals(topic, record.topic);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(key, value, timestamp, offset, topic);
+        }
+    }
+
+    private final static class Punctuation {
+        private final long intervalMs;
+        private final PunctuationType punctuationType;
+        private final Punctuator callback;
+
+        Punctuation(final long intervalMs,
+                    final PunctuationType punctuationType,
+                    final Punctuator callback) {
+            this.intervalMs = intervalMs;
+            this.punctuationType = punctuationType;
+            this.callback = callback;
+        }
+    }
+
+    private final class MockPunctuator implements Punctuator {
+        private final List<Long> punctuatedAt = new LinkedList<>();
+
+        @Override
+        public void punctuate(long timestamp) {
+            punctuatedAt.add(timestamp);
+        }
+    }
+
+    private final class MockProcessor implements Processor {
```

Comment:
We already have a `MockProcessor` in `o.a.k.test`, could we refactor that class and use it here? Similar for `MockProcessorSupplier`.